File: planetgen/noise.py
"""Seamless 3D gradient noise, fully vectorized in numpy.

Noise is evaluated at 3D points on (or near) the unit sphere, so there are no
map seams and no polar distortion. Everything is a pure function of position
and seed — the future game can re-evaluate the exact same fields at any
resolution for consistent local detail.
"""

from __future__ import annotations

import numpy as np

_MASK = np.int64(0xFFFFFFFF)

# fixed table of 256 well-distributed unit gradients
_rs = np.random.RandomState(1234567)
_g = _rs.normal(size=(256, 3))
_GRAD = _g / np.linalg.norm(_g, axis=1, keepdims=True)


def _hash(ix, iy, iz, seed):
    """Integer lattice hash -> [0, 255]. int64 arithmetic with 32-bit wrap."""
    h = (ix * np.int64(0x8DA6B343) + iy * np.int64(0xD8163841)
         + iz * np.int64(0xCB1AB31F) + np.int64(seed) * np.int64(0x9E3779B1)) & _MASK
    h = (h ^ (h >> 13)) * np.int64(0xC2B2AE35) & _MASK
    h = h ^ (h >> 16)
    return (h & 255).astype(np.intp)


def _fade(t):
    return t * t * t * (t * (t * 6.0 - 15.0) + 10.0)


def gradient_noise(p, seed=0):
    """Perlin-style gradient noise for points p (n, 3). Output roughly [-1, 1]."""
    pi = np.floor(p).astype(np.int64)
    pf = p - pi
    fx, fy, fz = _fade(pf[:, 0]), _fade(pf[:, 1]), _fade(pf[:, 2])
    total = np.zeros(len(p))
    for dx in (0, 1):
        wx = fx if dx else 1.0 - fx
        for dy in (0, 1):
            wy = fy if dy else 1.0 - fy
            for dz in (0, 1):
                wz = fz if dz else 1.0 - fz
                g = _GRAD[_hash(pi[:, 0] + dx, pi[:, 1] + dy, pi[:, 2] + dz, seed)]
                d = pf - np.array([dx, dy, dz], dtype=np.float64)
                total += (wx * wy * wz) * np.einsum("nd,nd->n", g, d)
    return total * 1.9


def fbm(p, octaves=5, freq=1.0, lacunarity=2.0, gain=0.5, seed=0):
    """Fractal Brownian motion; output roughly [-1, 1]."""
    total = np.zeros(len(p))
    amp, f, norm = 1.0, freq, 0.0
    for o in range(octaves):
        total += amp * gradient_noise(p * f, seed=seed * 7919 + o * 131)
        norm += amp
        amp *= gain
        f *= lacunarity
    return total / norm


def ridged(p, octaves=5, freq=1.0, lacunarity=2.0, gain=0.5, seed=0):
    """Ridged multifractal — sharp crests, good for young mountain texture. [0, 1]."""
    total = np.zeros(len(p))
    amp, f, norm = 1.0, freq, 0.0
    for o in range(octaves):
        n = 1.0 - np.abs(gradient_noise(p * f, seed=seed * 104729 + o * 131))
        total += amp * n * n
        norm += amp
        amp *= gain
        f *= lacunarity
    return total / norm


def fbm_vec3(p, octaves=4, freq=1.0, seed=0):
    """Three independent fBm channels -> (n, 3), for domain warping."""
    return np.stack([fbm(p, octaves, freq, seed=seed + 11),
                     fbm(p, octaves, freq, seed=seed + 23),
                     fbm(p, octaves, freq, seed=seed + 37)], 1)


def warp(p, amp=0.4, freq=1.0, octaves=4, seed=0, renormalize=True):
    """Domain-warped positions; renormalize keeps points on the unit sphere."""
    q = p + amp * fbm_vec3(p, octaves=octaves, freq=freq, seed=seed)
    if renormalize:
        q = q / np.linalg.norm(q, axis=1, keepdims=True)
    return q

File: viewer/src/noise.rs
//! Exact Rust port of planetgen/noise.py — same hash, same gradient table,
//! same octave scheme, so the same seed produces the same planet in both
//! worlds. Equality is enforced by tests/golden.rs against values generated
//! by the Python original (including its int64 wrap-around behavior, which
//! `wrapping_mul`/`wrapping_add` reproduce bit-exactly).

use crate::noise_grad::GRAD;
use glam::DVec3;

const MASK: i64 = 0xFFFF_FFFF;

#[inline]
fn hash(ix: i64, iy: i64, iz: i64, seed: i64) -> usize {
    let mut h = ix
        .wrapping_mul(0x8DA6_B343)
        .wrapping_add(iy.wrapping_mul(0xD816_3841))
        .wrapping_add(iz.wrapping_mul(0xCB1A_B31F))
        .wrapping_add(seed.wrapping_mul(0x9E37_79B1))
        & MASK;
    h = (h ^ (h >> 13)).wrapping_mul(0xC2B2_AE35) & MASK;
    h ^= h >> 16;
    (h & 255) as usize
}

#[inline]
fn fade(t: f64) -> f64 {
    t * t * t * (t * (t * 6.0 - 15.0) + 10.0)
}

/// A separate 64-bit lattice hash for procedural fields that need scalar
/// values rather than the legacy 256-entry gradient table. Keeping this out
/// of `hash` preserves the Python/Rust terrain-noise goldens byte-for-byte.
#[inline]
fn scalar_hash(ix: i64, iy: i64, iz: i64, seed: i64) -> u64 {
    let mut h = (ix as u64)
        .wrapping_mul(0x9E37_79B9_7F4A_7C15)
        ^ (iy as u64).wrapping_mul(0xC2B2_AE3D_27D4_EB4F)
        ^ (iz as u64).wrapping_mul(0x1656_67B1_9E37_79F9)
        ^ (seed as u64).wrapping_mul(0x85EB_CA77_C2B2_AE63);
    h ^= h >> 30;
    h = h.wrapping_mul(0xBF58_476D_1CE4_E5B9);
    h ^= h >> 27;
    h = h.wrapping_mul(0x94D0_49BB_1331_11EB);
    h ^ (h >> 31)
}

/// One cheap, approximately standard-normal lattice variate. The population
/// count of a well-mixed 64-bit word is Binomial(64, 1/2), with mean 32 and
/// variance 16. It is close to Gaussian and compiles to one popcount instead
/// of spending a dozen additions at every corner of every octave.
#[inline]
fn lattice_normal(bits: u64) -> f64 {
    (bits.count_ones() as f64 - 32.0) * 0.25
}

/// Seamless scalar value noise with an approximately N(0,1) marginal.
///
/// Trilinear interpolation normally has lower variance in the middle of a
/// lattice cell than at its corners. Dividing by the exact L2 norm of the
/// eight interpolation weights removes that phase-dependent variance. This
/// matters when a caller applies a CDF: every position should have the same
/// occupancy statistics, not just the lattice points.
pub fn normal_value_noise(p: DVec3, seed: i64) -> f64 {
    let pi = p.floor();
    let pf = p - pi;
    let (ix, iy, iz) = (pi.x as i64, pi.y as i64, pi.z as i64);
    let (fx, fy, fz) = (fade(pf.x), fade(pf.y), fade(pf.z));
    let mut total = 0.0;
    for dx in 0..2i64 {
        let wx = if dx == 1 { fx } else { 1.0 - fx };
        for dy in 0..2i64 {
            let wy = if dy == 1 { fy } else { 1.0 - fy };
            for dz in 0..2i64 {
                let wz = if dz == 1 { fz } else { 1.0 - fz };
                total += wx
                    * wy
                    * wz
                    * lattice_normal(scalar_hash(ix + dx, iy + dy, iz + dz, seed));
            }
        }
    }
    let norm2 = ((1.0 - fx).powi(2) + fx.powi(2))
        * ((1.0 - fy).powi(2) + fy.powi(2))
        * ((1.0 - fz).powi(2) + fz.powi(2));
    total / norm2.sqrt()
}

/// Perlin-style gradient noise at p; output roughly [-1, 1].
pub fn gradient_noise(p: DVec3, seed: i64) -> f64 {
    let pi = p.floor();
    let pf = p - pi;
    let (ix, iy, iz) = (pi.x as i64, pi.y as i64, pi.z as i64);
    let (fx, fy, fz) = (fade(pf.x), fade(pf.y), fade(pf.z));
    let mut total = 0.0;
    for dx in 0..2i64 {
        let wx = if dx == 1 { fx } else { 1.0 - fx };
        for dy in 0..2i64 {
            let wy = if dy == 1 { fy } else { 1.0 - fy };
            for dz in 0..2i64 {
                let wz = if dz == 1 { fz } else { 1.0 - fz };
                let g = GRAD[hash(ix + dx, iy + dy, iz + dz, seed)];
                let d = pf - DVec3::new(dx as f64, dy as f64, dz as f64);
                total += wx * wy * wz * (g[0] * d.x + g[1] * d.y + g[2] * d.z);
            }
        }
    }
    total * 1.9
}

pub fn fbm(p: DVec3, octaves: u32, freq: f64, seed: i64) -> f64 {
    let (mut total, mut amp, mut f, mut norm) = (0.0, 1.0, freq, 0.0);
    for o in 0..octaves as i64 {
        total += amp * gradient_noise(p * f, seed.wrapping_mul(7919).wrapping_add(o * 131));
        norm += amp;
        amp *= 0.5;
        f *= 2.0;
    }
    total / norm
}

pub fn ridged(p: DVec3, octaves: u32, freq: f64, seed: i64) -> f64 {
    let (mut total, mut amp, mut f, mut norm) = (0.0, 1.0, freq, 0.0);
    for o in 0..octaves as i64 {
        let n = 1.0
            - gradient_noise(p * f, seed.wrapping_mul(104_729).wrapping_add(o * 131)).abs();
        total += amp * n * n;
        norm += amp;
        amp *= 0.5;
        f *= 2.0;
    }
    total / norm
}

/// fbm evaluated only over octaves [first, first+count) of the full stack —
/// the per-LOD "band" trick: a quadtree node adds exactly the octaves its
/// resolution can carry, coarser nodes carry fewer. Same seeds per octave
/// index as `fbm`, so bands from different levels sum consistently.
pub fn fbm_band(p: DVec3, first: u32, count: u32, base_freq: f64, seed: i64) -> f64 {
    let mut total = 0.0;
    let mut amp = 0.5f64.powi(first as i32);
    let mut f = base_freq * 2.0f64.powi(first as i32);
    for o in first as i64..(first + count) as i64 {
        total += amp * gradient_noise(p * f, seed.wrapping_mul(7919).wrapping_add(o * 131));
        amp *= 0.5;
        f *= 2.0;
    }
    total
}

pub fn ridged_band(p: DVec3, first: u32, count: u32, base_freq: f64, seed: i64) -> f64 {
    let mut total = 0.0;
    let mut amp = 0.5f64.powi(first as i32);
    let mut f = base_freq * 2.0f64.powi(first as i32);
    for o in first as i64..(first + count) as i64 {
        let n = 1.0
            - gradient_noise(p * f, seed.wrapping_mul(104_729).wrapping_add(o * 131)).abs();
        total += amp * (n * n - 0.5);
        amp *= 0.5;
        f *= 2.0;
    }
    total
}
